middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -580,36 +647,45 @@ and let_cont_rec env res invariant_params conts body =
   let conts_to_handlers, res =
     Continuation.Map.fold
       (fun k handler (conts_to_handlers, res) ->
-        let vars, _arity, handler, res = continuation_handler env res handler in
+        let vars, _arity, handler, free_vars_of_handler, res =
+          continuation_handler env res handler
+        in
+        let free_vars =
+          C.remove_vars_with_machtype free_vars_of_handler invariant_vars
+        in
         ( Continuation.Map.add k
-            (invariant_vars @ vars, handler)
+            (invariant_vars @ vars, handler, free_vars)
             conts_to_handlers,
           res ))
       conts_to_handlers
       (Continuation.Map.empty, res)
   in
   let dbg = Debuginfo.none (* CR mshinwell: fix debuginfo *) in
+  let body, free_vars_of_body, res = expr env res body in
   (* Setup the Cmm handlers for the Ccatch *)
-  let handlers =
+  let handlers, free_vars =
     Continuation.Map.fold
-      (fun k (vars, handler) acc ->
+      (fun k (vars, handler, free_vars_of_handler) (handlers, free_vars) ->
+        let free_vars = Backend_var.Set.union free_vars free_vars_of_handler in
```

Comment:
`free_vars_of_handler` corresponds to the free variables of the handler minus the parameters. I think it's simpler if it corresponds to the free variables of the handler including the parameters, and we only remove them here, at the point they're actually bound.

Reply:
done